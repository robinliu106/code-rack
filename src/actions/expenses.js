import uuid from "uuid";
import database from "../firebase/firebase";

//ADD_SNIPPET
export const addSnippet = (expense) => ({
    type: "ADD_SNIPPET",
    expense,
});

export const startAddSnippet = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const { description = "", note = "", amount = 0, createdAt = 0 } = expenseData;
        const expense = { description, note, amount, createdAt };
        return database
            .ref(`users/${uid}/expenses`)
            .push(expense)
            .then((ref) => {
                dispatch(addSnippet({ id: ref.key, ...expense }));
            });
    };
};

//REMOVE_SNIPPET
export const removeSnippet = ({ id } = {}) => ({
    type: "REMOVE_SNIPPET",
    id,
});

export const startRemoveSnippet = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        //return async action
        return database
            .ref(`users/${uid}/expenses/${id}`)
            .remove()
            .then(() => {
                dispatch(removeSnippet({ id }));
            });
    };
};

//EDIT_SNIPPET
export const editSnippet = (id, updates) => ({
    type: "EDIT_SNIPPET",
    id,
    updates,
});

export const startEditSnippet = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database
            .ref(`users/${uid}/expenses/${id}`)
            .update(updates)
            .then(() => {
                dispatch(editSnippet(id, updates));
            });
    };
};

//SET_SNIPPETS
export const setExpenses = (expenses) => ({
    type: "SET_SNIPPETS",
    expenses,
});

// export const startSetSnippets;
export const startSetSnippets = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database
            .ref(`users/${uid}/expenses`)
            .once("value")
            .then((snapshot) => {
                const expenses = [];
                snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val(),
                    });
                });
                dispatch(setExpenses(expenses));
            });
    };
};
