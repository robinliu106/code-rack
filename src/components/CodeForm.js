import React, { useState } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

const now = moment();

const CodeForm = (props) => {
    const [title, setTitle] = useState(props.expense ? props.expense.description : "");
    const [note, setNote] = useState(props.expense ? props.expense.note : "");
    const [createdAt, setCreatedAt] = useState(props.expense ? moment(props.expense.createdAt) : moment());
    const [calendarFocused, setCalendarFocused] = useState(false);
    const [error, setError] = useState("");

    const onTitleChange = (e) => {
        setTitle(e.target.value);
    };
    const onNoteChange = (e) => {
        setNote(e.target.value);
    };

    const onDateChange = (createdAt) => {
        if (createdAt) {
            setCreatedAt(createdAt);
        }
    };
    const onFocusChange = ({ focused }) => {
        setCalendarFocused(focused);
    };
    const onSubmit = (e) => {
        e.preventDefault();
        if (!title) {
            setError("Please Provide a title");
        } else {
            setError("");

            props.onSubmit({
                description: title,
                createdAt: createdAt.valueOf(),
                note,
            });
        }
    };

    return (
        <form className="form" onSubmit={onSubmit}>
            {error && <p className="form__error">{error}</p>}
            <input
                className="text-input"
                type="text"
                placeholder="Title"
                autoFocus
                value={title}
                onChange={onTitleChange}
            />

            <SingleDatePicker
                date={createdAt}
                onDateChange={onDateChange}
                focused={calendarFocused}
                onFocusChange={onFocusChange}
                numberOfMonths={1}
                isOutsideRange={() => false}
            />
            <textarea className="textarea" value={note} onChange={onNoteChange}></textarea>
            <div>
                <button className="button">Save Code</button>
            </div>
        </form>
    );
};

export default CodeForm;
