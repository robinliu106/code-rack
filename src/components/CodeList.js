import React from "react";
import { connect } from "react-redux";
import CodeListItem from "./CodeListItem";
import selectSnippets from "../selectors/expenses";

export const CodeList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Snippets</div>
            <div className="show-for-desktop">Snippet</div>
            <div className="show-for-desktop">Description</div>
        </div>
        <div className="list-body">
            {props.expenses.length === 0 ? (
                <div className="list-item list-item--message">
                    <span>No Snippets</span>
                </div>
            ) : (
                props.expenses.map((expense) => {
                    return <CodeListItem key={expense.id} {...expense} />;
                })
            )}
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectSnippets(state.expenses, state.filters),
    };
};

export default connect(mapStateToProps)(CodeList);
