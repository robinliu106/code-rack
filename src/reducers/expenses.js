//Expenses Reducer
const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case "ADD_SNIPPET":
            return [...state, action.expense];
        case "REMOVE_SNIPPET":
            return state.filter(({ id }) => action.id !== id);
        case "EDIT_SNIPPET":
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates, //override amount (passed in through action.updates)
                    };
                } else {
                    return expense;
                }
            });
        case "SET_SNIPPETS":
            return action.expenses;

        default:
            return state;
    }
};
