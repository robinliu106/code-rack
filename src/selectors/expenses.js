import moment from "moment";

//Get visible expenses
export default (expenses, { text }) => {
    return expenses
        .filter((expense) => {
            const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
            return textMatch;
        })
        .sort((a, b) => {
            return a.createdAt < b.createdAt ? 1 : -1;
        });
};
