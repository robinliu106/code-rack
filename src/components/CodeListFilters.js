import React, { useState } from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from "../actions/filters";

const CodeListFilters = (props) => {
    const onTextChange = (e) => {
        props.setTextFilter(e.target.value);
    };

    return (
        <div className="content-container">
            <div className="input-group">
                <div className="input-group__item">
                    <input type="text" className="text-input" onChange={onTextChange} placeholder="Search Code" />
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => ({
    filters: state.filters,
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
});

export default connect(mapStateToProps, mapDispatchToProps)(CodeListFilters);
