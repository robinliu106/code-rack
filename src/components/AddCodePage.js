import React from "react";
import { connect } from "react-redux";
import CodeForm from "./CodeForm";
import { startAddExpense } from "../actions/expenses";

const AddSnippetPage = (props) => {
    const onSubmit = (expense) => {
        props.startAddExpense(expense);
        props.history.push("/"); //redirect to dashboard upon submit
    };

    return (
        <div>
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Add Code</h1>
                </div>
            </div>
            <div className="content-container">
                <CodeForm onSubmit={onSubmit} />
            </div>
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(AddSnippetPage);
