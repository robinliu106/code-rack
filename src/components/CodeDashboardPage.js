import React from "react";
import CodeListFilters from "./CodeListFilters";
import CodeSummary from "./CodeSummary";
import CodeList from "./CodeList";

const CodeDashboardPage = () => (
    <div>
        <CodeSummary />
        <CodeListFilters />
        <CodeList />
    </div>
);

export default CodeDashboardPage;
