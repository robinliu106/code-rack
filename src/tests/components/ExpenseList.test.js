import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

test("Should render ExpenseList with expenses", () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot(); //first run never fails, creates a snapshot
});

test("Should render ExpenseList with empty message", () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});
