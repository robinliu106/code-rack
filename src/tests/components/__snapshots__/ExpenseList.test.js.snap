// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseList with empty message 1`] = `
"<div>
  <h1>
    Expense List
  </h1>
  <p>
    No Expenses
  </p>
</div>"
`;

exports[`Should render ExpenseList with expenses 1`] = `
"<div>
  <h1>
    Expense List
  </h1>
  <ExpenseListItem id=\\"1\\" description=\\"Gum\\" note=\\"\\" amount={195} createdAt={0} />
  <ExpenseListItem id=\\"2\\" description=\\"rent\\" note=\\"\\" amount={300095} createdAt={-345600000} />
  <ExpenseListItem id=\\"3\\" description=\\"credit card\\" note=\\"\\" amount={6700} createdAt={345600000} />
</div>"
`;
